lib/pengine/unpack.c

```diff
@@ -908,6 +908,82 @@ unpack_resources(const xmlNode *xml_resources, pcmk_scheduler_t *scheduler)
     return TRUE;
 }
 
+/*!
+ * \internal
+ * \brief Unpack essential information from topology request XML
+ *
+ * \param[in]  xml     Request XML to search
+ * \param[out] id      If not NULL, where to store level number
+ *
+ * \return Topology level XML from within \p xml, or NULL if not found
+ * \note The caller is responsible for freeing \p *target and \p *desc if set.
+ */
+static xmlNode *
+unpack_level_request(xmlNode *xml, int *id)
+{
+    int local_id = 0;
+
+    /* The level element can be the top element or lower. If top level, don't
+     * search by xpath, because it might give multiple hits if the XML is the
+     * CIB.
+     */
+    if ((xml != NULL) && !pcmk__xe_is(xml, PCMK_XE_FENCING_LEVEL)) {
+        xml = get_xpath_object("//" PCMK_XE_FENCING_LEVEL, xml, LOG_WARNING);
+    }
```

Comment:
I just realized that this looks wrong in the fencer. `get_xpath_object()` always searches the entire XML document that `xml` is part of, not just the subtree starting with `xml`. I'm not sure how it's working there. :-/ In any case, here we *know* it's the CIB we're unpacking, so `<fencing-level>` has to be the immediate child of `<fencing-topology>`.

In this function, we can assume `xml` is non-NULL and a `PCMK_XE_FENCING_LEVEL` element. And at that point, it doesn't really need to be a function.